peps/pep-0781.rst

```diff
@@ -0,0 +1,280 @@
+PyPI. To avoid breaking users of one of the existing bindings, this PEP
+proposes introducing a new namespace for compression libraries,
+``compression``. This name is already reserved on PyPI for use in the
+standard library. The new Zstandard module will be ``compression.zstd``.
+Other compression modules will be re-exported to the ``compression`` namespace
+and their current import names will be deprecated.
+
+Providing a common namespace for compression modules has several advantages.
+First, it reduces user confusion about where to find compression modules.
+Second, the top level ``compression`` module could provide information on which
+compression formats are available, similar to ``hashlib``'s
+``algorithms_available``. If :pep:`775` is accepted, a
+``compression.algorithms_guaranteed`` could be provided as well, listing
+``zlib``. Finally, a ``compression`` namespace prevents future issues with
+merging other compression formats into the standard library. New compression
+formats will likely be published to PyPI prior to integration into
+CPython. Therefore, any new compression format import name will likely already
+be claimed by the time a module would be considered for inclusion in CPython.
+Putting compression modules under a package prefix prevents issues with
+potential future name clashes.
+
+Specification
+=============
+
+The ``compression`` namespace
+-----------------------------
+
+A new namespace package for compression modules will be introduced named
+``compression``. The top-level module for this package will be empty to begin
+with, but a standard API for interacting with compression routines may be
+added in the future to the toplevel.
+
+The ``compression.zstd`` module
+-------------------------------
+
+A new module, ``compression.zstd`` will be introduced with Zstandard
+compression APIs that match other compression modules in the standard library,
+namely
+
+* ``compress`` / ``decompress`` - APIs for one-shot compression/decompression
+* ``ZstdFile`` / ``open`` - APIs for interacting with streams and file-like
+  objects
+* ``ZstdCompressor`` / ``ZstdDecompressor`` - APIs for incremental compression/
+  decompression
+
+It will also contain some Zstandard-specific functionality
+
+* ``ZstdDict`` / ``train_dict`` / ``finalize_dict`` - APIs for interacting with
+  Zstandard dictionaries, which are useful for compressing many small chunks of
+  similar data
+
+Other compression modules
+-------------------------
+
+New import names ``compression.lzma``, ``compression.bz2``, and
+``compression.zlib`` will be introduced for the existing standard library
+compression modules ``lzma``, ``bz2``, and ``zlib`` respectively. The new
+modules will simply re-export the contents of the existing modules. Importing
+the existing module import names will emit a deprecation warning, with a
+planned removal in 3.24. The documentation for these modules will be updated
+to discuss the planned deprecation and removal.
```

Comment:
I am not sure if it is more readable, but here is an alternative formulation:

Saying that:
- The modules `lzma`/`bz2`/`zlib` will be “renamed” to `compression.lzma`/`compression.bz2`/`compression.zlib`
- A temporary implementation of modules `lzma`/`bz2`/`zlib` will simply re-export the corresponding `compression` sub-module while emitting a deprecation warning
- The modules `lzma`/`bz2`/`zlib` are planned for removal in 3.24
- Documentation will be updated accordingly

Up to you!

Reply:
The tricky thing is that the actual module code is *not* moving yet. At least, I don't think that is totally safe since things could rely on the name of the module they are contained in.

For now, I was going to have `from zlib import *` in the `compression` sub-modules. In the future that would be swapped. But I agree my current explanation is not clear at all :)

Reply:
I re-wrote this section, hopefully it makes more sense. I also added a small note about how to use these changes in a compatible manner to the motivation section.